engine/security: add more description about ssh://

AkihiroSuda opened the pull request:
<!--Thanks for your contribution. See [CONTRIBUTING](CONTRIBUTING.md)
    for this project's contribution guidelines. Remove these comments
    as you go.

    Help us merge your changes more quickly by adding details and setting metadata
    (such as labels, milestones, and reviewers) over at the right-hand side.-->

### Proposed changes

<!--Tell us what you did and why-->

Previously, the `ssh://` helper was only mentioned in `engine/security/index.md`.

The `ssh://` helper is now documented in "Protect the Docker daemon socket" (`engine/security/protect-access.md`, nee `engine/security/https.md`).

AkihiroSuda commented:
cc @tonistiigi

thaJeztah commented on engine/install/linux-postinstall.md:
```diff
@@ -151,7 +151,7 @@ the [Docker CLI Reference](/engine/reference/commandline/dockerd/) article.
 > it is possible for remote non-root users to gain root access on the host. For more information on how to use TLS 
 > certificates to secure this connection, check this article on 
-> [how to protect the Docker daemon socket](../security/https.md).
+> [how to protect the Docker daemon socket](../security/protect-socket.md).
```
Thinking out loud; would `secure-api`, `protect-access` or something along those lines be more descriptive? (as it's not necessarily the "socket" we're protecting, but protecting access to the engine API.

Perhaps we should move this warning to an include-file, so that we can more easily re-use it (thinking of adding the same / similar warning on https://docs.docker.com/engine/api/)

AkihiroSuda replied:
Changed to protect-access

thaJeztah commented on engine/security/protect-socket.md:
```diff
@@ -1,17 +1,41 @@
+## Use SSH to protect the Docker daemon socket
+
+The `docker` CLI can talk to a remote Docker daemon via SSH,
+by specifying `$DOCKER_HOST` to `ssh://<USERNAME>@<HOST>`.
```
Perhaps would be good to;

- use a different username than `root` as example (to encourage setting up a user for docker; https://docs.docker.com/engine/install/linux-postinstall/)
- should we describe that `username` is optional (as it can be configured in `~/.ssh/config`) ?
- _slightly_ prefer (e.g.) `host1.example.com` to prevent confusing the above from being an e-mail address (`root@example.com` would be easily confused to be an e-mail address)
- I think we should describe using `docker context` as the "main" method to do this (assuming this is not for an "ad-hoc" situation)

AkihiroSuda replied:
updated

thaJeztah commented on engine/security/protect-socket.md:
````diff
@@ -1,17 +1,41 @@
+$ docker ps
+```
+
+The support for SSH is available since Docker Engine 18.09.
+
````
I think we can remove this, as 18.09 is EOL, so any "current" version of docker now supports it (people shouldn't be using older versions as they're insecure now)

```suggestion
```

AkihiroSuda replied:
updated

thaJeztah commented on engine/security/protect-socket.md:
````diff
@@ -1,17 +1,41 @@
+The `docker` CLI can talk to a remote Docker daemon via SSH,
+by specifying `$DOCKER_HOST` to `ssh://<USERNAME>@<HOST>`.
+
+```console
+$ export DOCKER_HOST=ssh://root@example.com
+$ docker ps
+```
````
Perhaps something like this:

> **Note**
>
> The given `USERNAME` must have permissions to access the docker socket on the
> remote machine. Refer to [manage Docker as a non-root user](../../install/linux-postinstall.md#manage-docker-as-a-non-root-user)
> to learn how to give a non-root user access to the docker socket.

The following example creates a [`docker context`](../../context/working-with-contexts.md)
to connect with a remote `dockerd` daemon on `host1.example.com` using SSH, and
as the `docker-user` user on the remote machine:

```console
$ docker context create \
    --docker host=ssh://docker-user@host1.example.com \
    --description="Remote engine" \
    my-remote-engine

my-remote-engine
Successfully created context "my-remote-engine"
```

After creating the context, use `docker context use` to switch the `docker` CLI
to use it, and to connect to the remote engine:

```console
$ docker context use my-remote-engine
my-remote-engine
Current context is now "my-remote-engine"

$ docker info
<prints output of the remote engine>
```

Use use the `default` context to switch back to the default (local) daemon:

```console
$ docker context use default
default
Current context is now "default"
```

Alternatively, use the `DOCKER_HOST` environment variable to temporarily switch
the `docker` CLI to connect to the remote host using SSH. This does not require
creating a context, and can be useful to create an ad-hoc connection with a different
engine:

```console
$ export DOCKER_HOST=ssh://docker-user@host1.example.com
$ docker info
<prints output of the remote engine>
```

thaJeztah replied:
Markdown for the above (GitHub "suggestions" didn't like the nested code-fences 😞)

    > **Note**
    >
    > The given `USERNAME` must have permissions to access the docker socket on the
    > remote machine. Refer to [manage Docker as a non-root user](../../install/linux-postinstall.md#manage-docker-as-a-non-root-user)
    > to learn how to give a non-root user access to the docker socket.

    The following example creates a [`docker context`](../../context/working-with-contexts.md)
    to connect with a remote `dockerd` daemon on `host1.example.com` using SSH, and
    as the `docker-user` user on the remote machine:

    ```console
    $ docker context create \
        --docker host=ssh://docker-user@host1.example.com \
        --description="Remote engine" \
        my-remote-engine

    my-remote-engine
    Successfully created context "my-remote-engine"
    ```

    After creating the context, use `docker context use` to switch the `docker` CLI
    to use it, and to connect to the remote engine:

    ```console
    $ docker context use my-remote-engine
    my-remote-engine
    Current context is now "my-remote-engine"

    $ docker info
    <prints output of the remote engine>
    ```

    Use use the `default` context to switch back to the default (local) daemon:

    ```console
    $ docker context use default
    default
    Current context is now "default"
    ```

    Alternatively, use the `DOCKER_HOST` environment variable to temporarily switch
    the `docker` CLI to connect to the remote host using SSH. This does not require
    creating a context, and can be useful to create an ad-hoc connection with a different
    engine:

    ```console
    $ export DOCKER_HOST=ssh://docker-user@host1.example.com
    $ docker info
    <prints output of the remote engine>
    ```

AkihiroSuda replied:
updated

thaJeztah reviewed (changes requested):
Thanks! I left some thoughts/suggestions.

Wondering if the Dockerfile and shell-scripts are still used anywhere 🤔 (if they're used by external people, we may need to check if that's a problem, as we won't have redirects for them)

AkihiroSuda commented:
Thanks, updated

> Wondering if the Dockerfile and shell-scripts are still used anywhere 🤔 (if they're used by external people, we may need to check if that's a problem, as we won't have redirects for them)

Reverted Dockerfile and scripts

thaJeztah commented:
Oh! found one more occurrence of the old URL in the TOC; https://github.com/docker/docker.github.io/blob/60624cfb4a44268db19715a9ce6cf6c4a3b112f9/_data/toc.yaml#L208

Other than that, this looks great now, thanks! (Guess it can be moved out of draft?)

AkihiroSuda commented:
Thanks, fixed

AkihiroSuda commented:
Preview is here: https://deploy-preview-12237--docsdocker.netlify.app/engine/security/protect-access/

thaJeztah reviewed (approved):
LGTM!

This is a great enhancement, thank you!

Reading the page, there's a couple of other things that I think could use some enhancement, but adding that to my list to work on if I find time 😅
